### ISSUE TITLE
[Flaky test] [sig-network] Services should be able to change the type and ports of a service [Slow]

### ISSUE THREAD

**Issue by marseel** [kubernetes/kubernetes#98099]
**Which jobs are flaking**:
gce-master-scale-correctness
**Which test(s) are flaking**:
Services should be able to change the type and ports of a service [Slow]
**Testgrid link**:
https://k8s-testgrid.appspot.com/sig-scalability-gce#gce-master-scale-correctness
**Reason for failure**:
Test timeouts in ~half of the runs since it was enabled here: https://github.com/kubernetes/kubernetes/pull/97345

/sig scalability
/sig network

**Comment by wojtek-t** [comment]
@kubernetes/sig-network-bugs 

@kubernetes/sig-scalability

**Comment by wojtek-t** [comment]
The context is - the problems are strictly related to scale - 5k-node jobs are timing out.


I very briefly skimmed through the test and didn't see anything obviously suspicious (apart from the fact there are gazillions long timeouts there, which alltogether sum up long hours).

That said - it seems to me that there are effectively two separate tests inside it:
- one for TCP
- one for UDP
They don't seem to share anything with each other.

@kubernetes/sig-network-bugs @thockin @aojea - would it make sense to split this test two effectively two separate tests: one for TCP and one for UDP. (I guess it would also allow us to share some code there).

**Comment by wojtek-t** [comment]
[And potentially it could help with debugging the problem if it is strictly related to only one of those cases...]

**Comment by aojea** [comment]
> The context is - the problems are strictly related to scale - 5k-node jobs are timing out.

I wanted to focus on this past week but I didn´t have time, this is my list of more expensive times and approx average duration, all of the sig-network ones use LoadBalancers. I wanted to time the test to understand if there are places where can save time ...

```
Kubernetes e2e suite.[sig-network] Services [Slow] should be able to create and tear down a standard-tier load balancer [Slow] 60 mins
Kubernetes e2e suite.[sig-network] Services should be able to change the type and ports of a service [Slow] 60 mins
Kubernetes e2e suite.[sig-network] Services should be able to create an internal type load balancer [Slow] 40 mins
Kubernetes e2e suite.[sig-network] ESIPP [Slow] should work for type=LoadBalancer 30 mins
Kubernetes e2e suite.[sig-network] Loadbalancing: L7 [Slow] Nginx should conform to Ingress spec 30 mins
Kubernetes e2e suite.[sig-network] Services should only allow access from service loadbalancer source ranges [Slow] 20 mins
Kubernetes e2e suite.[sig-network] ESIPP [Slow] should work from pods 20 mins
Kubernetes e2e suite.[k8s.io] Pods should cap back-off at MaxContainerBackOff [Slow][NodeConformance] 25 mins
Kubernetes e2e suite.[sig-network] ESIPP [Slow] should only target nodes with endpoints 25 mmins
```

> would it make sense to split this test two effectively two separate tests: one for TCP and one for UDP. (I guess it would also allow us to share some code there).

I don´t have much context on those tests, maybe Tim can help us more, but this comment makes me think that the use of both protocols is intended
https://github.com/kubernetes/kubernetes/blob/a410c14020017c7bedeb4c501758088155c66fb7/test/e2e/network/service.go#L1254-L1257

**Comment by thockin** [comment]
/triage accepted

It's been a very long time since I touched this, and I recall holding my nose on it in the first place.  Splitting TCP from UDP seems reasonable if we can run them in parallel.  we'll want to share the flow, and parameterize some shared funcs.

I don't know that this will prevent test timeouts, though.

**Comment by thockin** [comment]
We could use a volunteer to go poke at this, and see how hard it will really be.

**Comment by wojtek-t** [comment]
>  Splitting TCP from UDP seems reasonable if we can run them in parallel.

You mean "forcing them to run in parallel"? Or "allowing them to run in parallel"?

The later doesn't require any additiona work. The former I think mean that it will still have to be a single test (just the code will be organized better) and that won't really help with debugging (nor preventing timeouts).

**Comment by aojea** [comment]
this was discussed yesterday in sig-net meeting and my understanding was to split current test in 2 different so the test can run in parallel, if you want to do so and one protocol doesn´t have to wait for the other.
I think that is what you meant here

> would it make sense to split this test two effectively two separate tests: one for TCP and one for UDP. (I guess it would also allow us to share some code there).

**Comment by thockin** [comment]
As long as we CAN parallelize.  Creating LBs can be sloooooow.  :P

**Comment by wojtek-t** [comment]
Reopening - I don't think just splitting will address the problem. It may help us with debugging though...

**Comment by aojea** [comment]
With the split we moved from >60 mins to run the test, to run 2 tests in [30-40] mins each
https://k8s-testgrid.appspot.com/sig-release-master-informing#gce-master-scale-correctness&include-filter-by-regex=should%20be%20able&width=20&graph-metrics=test-duration-minutes
Seems a win :smile:

**Comment by wojtek-t** [comment]
Awesome - that's something I was silently counting for TBH :)
[I doubt it will fix flakiness, but maybe it will make it easier to debug...]

**Comment by wojtek-t** [comment]
Actually - I think we're getting close thanks to the two changes (https://github.com/kubernetes/kubernetes/pull/98181 + this one).

Here is the run:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-correctness/1354036770185940992

This particular one filed on UDP (not sure if it's just UDP-related, or just manifested with UDP now).
And we have a clear failure now:

```
Kubernetes e2e suite: [sig-network] LoadBalancers should be able to change the type and ports of a UDP service [Slow] 

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/network/loadbalancer.go:284 Jan 26 13:19:17.628: Could not reach UDP service through 34.74.118.37:80 after 45m0s: timed out waiting for the condition /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/network/loadbalancer.go:461
```

In my PR I decreased the timeout from 2h to 45m, as it seemed to me that in the past it should be enough.
But it failed on 45m timeout here.

So I think the next question now is:
- is it that slow - that we need more than 45m
or
- something isn't working as expected (either in the k8s itself or in the test)

But I think we have a clear single place to look at now.

**Comment by wojtek-t** [comment]
Actually - the full stacktrace:
```
Full Stack Trace
k8s.io/kubernetes/test/e2e/network.glob..func16.4()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/network/loadbalancer.go:461 +0x18bc
k8s.io/kubernetes/test/e2e.RunE2ETests(0xc002f38480)
	_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e.go:130 +0x36c
k8s.io/kubernetes/test/e2e.TestE2E(0xc002f38480)
	_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e_test.go:144 +0x2b
testing.tRunner(0xc002f38480, 0x4f18258)
	/usr/local/go/src/testing/testing.go:1123 +0xef
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1168 +0x2b3
```

So it's this line that failed:
https://github.com/kubernetes/kubernetes/blob/a38b7e5ecbf2aa3c960957235024ac7c3a444103/test/e2e/network/loadbalancer.go#L463

What is interesting is that something was changing during those 45 minutes:
```
Jan 26 12:34:17.533: INFO: Poking udp://34.74.118.37:80
Jan 26 12:34:17.565: INFO: Poke("udp://34.74.118.37:80"): read udp 10.12.5.206:58816->34.74.118.37:80: read: connection refused
Jan 26 12:34:19.565: INFO: Poking udp://34.74.118.37:80
Jan 26 12:34:19.598: INFO: Poke("udp://34.74.118.37:80"): read udp 10.12.5.206:50301->34.74.118.37:80: read: connection refused
...
Jan 26 12:38:35.565: INFO: Poking udp://34.74.118.37:80
Jan 26 12:38:35.598: INFO: Poke("udp://34.74.118.37:80"): read udp 10.12.5.206:56145->34.74.118.37:80: read: connection refused
Jan 26 12:38:37.565: INFO: Poking udp://34.74.118.37:80
Jan 26 12:38:40.566: INFO: Poke("udp://34.74.118.37:80"): read udp 10.12.5.206:49385->34.74.118.37:80: i/o timeout
Jan 26 12:38:41.565: INFO: Poking udp://34.74.118.37:80
Jan 26 12:38:44.565: INFO: Poke("udp://34.74.118.37:80"): read udp 10.12.5.206:57074->34.74.118.37:80: i/o timeout
Jan 26 12:38:45.565: INFO: Poking udp://34.74.118.37:80
...
Jan 26 12:45:25.565: INFO: Poking udp://34.74.118.37:80
Jan 26 12:45:28.565: INFO: Poke("udp://34.74.118.37:80"): read udp 10.12.5.206:33638->34.74.118.37:80: i/o timeout
Jan 26 12:45:29.565: INFO: Poking udp://34.74.118.37:80
Jan 26 12:45:29.597: INFO: Poke("udp://34.74.118.37:80"): read udp 10.12.5.206:48181->34.74.118.37:80: read: connection refused
Jan 26 12:45:31.565: INFO: Poking udp://34.74.118.37:80
Jan 26 12:45:31.598: INFO: Poke("udp://34.74.118.37:80"): read udp 10.12.5.206:47457->34.74.118.37:80: read: connection refused
...
Jan 26 13:19:17.597: INFO: Poking udp://34.74.118.37:80
Jan 26 13:19:17.628: INFO: Poke("udp://34.74.118.37:80"): read udp 10.12.5.206:37702->34.74.118.37:80: read: connection refused
Jan 26 13:19:17.628: FAIL: Could not reach UDP service through 34.74.118.37:80 after 45m0s: timed out waiting for the condition
```

**Comment by wojtek-t** [comment]
Note that it's 4th time we actually wait for LB in this test.

The failure happens after the pod were scaled down to 0 and then to 1.

So that makes me think, that's it's actually not LB-related, but rather something internally in k8s, becuase L4 LB doesn't depend on pods - it targets nodes...

**Comment by aojea** [comment]
If it is always UDP, I start to see ghosts between GCE and UDP ... we only have KIND to compare against, or do we have some other non gce enviroment?

**Comment by thockin** [comment]
@bowei @freehan - this does not look "normal".  Can we give some prio to helping track this flake?

**Comment by bowei** [comment]
hmm interesting. we will investigate.

**Comment by wojtek-t** [comment]
@bowei - any progress on this?

**Comment by swetharepakula** [comment]
I haven't had a chance to look into this yet. I will prioritize it this week.

**Comment by swetharepakula** [comment]
Digging through the logs I believe this was a bug in the test code that was fixed by #98405.

In the test, the service port changes around 12:33. However due to the bug, the test continues to use the old service port as seen in the test output:
```
 Jan 26 12:33:56.840: INFO: service port UDP: 80
```
The port should be 81. 
I believe the following sequence occurs from this point:
1) The reachability test fails in the beginning because the service has changed the port and the load balancer has not been updated yet resulting in the `read: connection refused` error. 
1) Around 12:38, the service controller is syncing the service and notices that the service port has changed and starts to delete the forwarding rule causing the error to change to `i/o timeout`. 
1) Once the forwarding rule is recreated, the error goes back to being `read: connection refused` because the forwarding rule exists but the wrong port is used.
```

I0126 12:38:33.132606      11 gce_loadbalancer_external.go:753] LoadBalancer port range for forwarding rule a66e04aea5b884ea8829bf2036bd1c32 was expected to be 80-80, but was actually 81-81

I0126 12:45:18.761987      11 gce_loadbalancer_external.go:254] ensureExternalLoadBalancer(a66e04aea5b884ea8829bf2036bd1c32(loadbalancers-2758/mutability-test)): Deleted forwarding rule.

I0126 12:52:24.753379      11 gce_loadbalancer_external.go:271] ensureExternalLoadBalancer(a66e04aea5b884ea8829bf2036bd1c32(loadbalancers-2758/mutability-test)): Created forwarding rule, IP 34.74.118.37.
```

Looking at [testgrid](https://testgrid.k8s.io/google-gce#gce-master-scale-correctness), the test runs afterward the one linked have all passed. Is there anything else that needs to be investigated as part of this issue?

**Comment by wojtek-t** [comment]
OK - the test was quite flaky in the past but indeed it didn't flake for the last 3 weeks (neither this nor TCP one).

I'm fine with closing this one - we can open a new one if this happens again.

@swetharepakula - thanks for debugging.


@aojea - it seems that the splitting of the test actually made things waay better. Which is great.
